crates/sui-core/src/node_sync/node_follower.rs

```diff
@@ -2,797 +2,538 @@
-            Ok(r) => r,
-        };
-
-        (
-            tokio::spawn(async move { state.handle_messages(&mut receiver).await }),
-            sender,
-        )
-    }
+    debug!(?peer, "follow_one_peer");
 
-    async fn handle_messages(self: Arc<Self>, receiver: &mut mpsc::Receiver<DigestsMessage>) {
-        // this pattern for limiting concurrency is from
-        // https://github.com/tokio-rs/tokio/discussions/2648
-        let limit = Arc::new(Semaphore::new(MAX_NODE_SYNC_CONCURRENCY));
+    // Global timeout, we do not exceed this time in this task.
+    let mut results = FuturesUnordered::new();
 
-        while let Some(DigestsMessage { sync_arg, tx }) = receiver.recv().await {
-            let state = self.clone();
-            let limit = limit.clone();
-
-            // hold semaphore permit until task completes. unwrap ok because we never close
-            // the semaphore in this context.
-            let permit = limit.acquire_owned().await.unwrap();
-
-            tokio::spawn(async move {
-                let res = state.process_digest(sync_arg, permit).await;
-                if let Err(error) = &res {
-                    error!(?sync_arg, "process_digest failed: {}", error);
-                }
+    let result_block = |fut, seq, digests| async move {
+        fut.await?;
+        trace!(?peer, ?seq, ?digests, "digest handler finished");
+        Ok::<TxSequenceNumber, SuiError>(seq)
+    };
 
-                // Notify waiters even if tx failed, to avoid leaking resources.
-                let digest = sync_arg.transaction_digest();
-                trace!(?digest, "notifying waiters");
-                state
-                    .pending_txes
-                    .notify(digest, ())
-                    .await
-                    .tap_err(|e| debug!(?digest, "{}", e))
-                    .ok();
-
-                if let Some(tx) = tx {
-                    // Send status back to follower so that it knows whether to advance
-                    // the watermark.
-                    if tx.send(res).is_err() {
-                        // This will happen any time the follower times out and restarts, but
-                        // that's ok - the follower won't have marked this digest as processed so it
-                        // will be retried.
-                        debug!(
-                            ?sync_arg,
-                            "could not send process_digest response to caller",
-                        );
-                    }
-                }
-            });
-        }
+    // Using a macro to avoid duplicating code - was much too difficult to satisfy the borrow
+    // checker with closures.
+    macro_rules! process_digest {
+        ($seq: expr, $digests: expr) => {{
+            let seq = $seq;
+            let digests = $digests;
+            let fut = sync_handle.handle_digest(peer, seq, digests).await?;
+            results.push(result_block(fut, seq, digests));
+        }};
```

Comment:
this invokes `handle_digest` and pushes the result future into the `results` queue.

I'm sorry about the macro but I couldn't find a way to avoid duplicating this code with functions, due to the borrow checker.